zerver/lib/user_groups.py

```diff
@@ -774,6 +774,30 @@ def get_recursive_subgroups_for_groups(
     return recursive_subgroups
 
 
+def get_root_id_annotated_recursive_subgroups_for_groups(
+    user_group_ids: Iterable[int], realm_id: int
+) -> QuerySet[NamedUserGroup]:
+    # Same as get_recursive_subgroups_for_groups but keeps track of
+    # each group root_id and annotates it.
```

Comment:
annotates it with what?

Reply:
which each group, i.e with each object of the fetched result

Reply:
I think this might not do what we want; see https://github.com/zulip/zulip/pull/33097#discussion_r1927963914 for details, but I don't see how this can possibly handle the possibility of multiple groups correctly. 

Probably the test for this code needs to be extended to exercise that case -- make two groups that each contain the same child group, and try to mention both of them in a message.